src/main/scala/com/cloudera/sparkts/ARIMA.scala

```diff
@@ -0,0 +1,566 @@
+      }
+    })
+    val gradient = new ObjectiveFunctionGradient(new MultivariateVectorFunction() {
+      def value(params: Array[Double]): Array[Double] = {
+        new ARIMAModel(order, params, includeIntercept).gradientlogLikelihoodCSSARMA(diffedY)
+      }
+    })
+    val initialGuess = new InitialGuess(initParams)
+    val maxIter = new MaxIter(10000)
+    val maxEval = new MaxEval(10000)
+    val goal = GoalType.MAXIMIZE
+    val optimal = optimizer.optimize(objFunction, gradient, goal, initialGuess, maxIter, maxEval)
+    optimal.getPoint
+  }
+
+
+  // TODO: implement MLE parameter estimates with Kalman filter.
+  private def fitWithML(
+      diffedY: Array[Double],
+      arTerms: Array[Array[Double]],
+      maTerms: Array[Double],
+      initCoeffs: Array[Double]): Array[Double] = {
+    throw new UnsupportedOperationException()
+  }
+
+  /**
+   * Initializes ARMA parameter estimates using the Hannan-Risannen algorithm. The process is:
+   * fit an AR(m) model of a higher order (i.e. m > max(p, q)), use this to estimate errors,
+   * then fit an OLS model of AR(p) terms and MA(q) terms. The coefficients estimated by
+   * the OLS model are returned as initial parameter estimates
+   * @param diffedY differenced time series, as appropriate
+   * @param order model order (p, d, q)
+   * @param includeIntercept boolean indicating if an intercept should be fit as well
+   * @return initial ARMA(p, d, q) parameter estimates
+   */
+  private def HannanRisannenInit(
+      diffedY: Array[Double],
+      order:(Int, Int, Int),
+      includeIntercept: Boolean): Array[Double] = {
+    val (p, d, q) = order
+    val addToLag = 1
+    val m = math.max(p, q) + addToLag // m > max(p, q)
+    // higher order AR(m) model
+    val arModel = Autoregression.fitModel(new DenseVector(diffedY), m)
+    val arTerms1 = Lag.lagMatTrimBoth(diffedY, m, includeOriginal = false)
+    val yTrunc = diffedY.drop(m)
+    val estimated = arTerms1.zip(
+      Array.fill(yTrunc.length)(arModel.coefficients)
+      ).map { case (v, b) => v.zip(b).map { case (yi, bi) => yi * bi}.sum + arModel.c }
+    // errors estimated from AR(m)
+    val errors = yTrunc.zip(estimated).map { case (y, yhat) => y - yhat }
+    // secondary regression, regresses X_t on AR and MA terms
+    val arTerms2 = Lag.lagMatTrimBoth(yTrunc, p, includeOriginal = false).drop(math.max(q - p, 0))
+    val errorTerms = Lag.lagMatTrimBoth(errors, q, includeOriginal = false).drop(math.max(p - q, 0))
+    val allTerms = arTerms2.zip(errorTerms).map { case (ar, ma) => ar ++ ma }
+    val regression = new OLSMultipleLinearRegression()
+    regression.setNoIntercept(!includeIntercept)
+    regression.newSampleData(yTrunc.drop(m - addToLag), allTerms)
+    val params = regression.estimateRegressionParameters()
+    params
+  }
+
+  /**
+   * Calculate a differenced vector of a given order. Size-preserving by leaving first `order`
```

Comment:
`com.cloudera.sparkts.TimeSeries` has a differences method, but this one is meant to work on `Vector`. It is also size-preserving, which is different from the one in `TimeSeries`. I've moved over to `UnivariateTimeSeries`. I agree that it makes more sense to have it there.